src/storage/index/index_manager.cpp

```diff
@@ -0,0 +1,189 @@
+  catalog::SqlTableHelper *sql_table_helper = catalog->GetUserTable(txn1, db_oid, ns_oid, table_oid);
+  // user table does not exist
+  if (sql_table_helper == nullptr) {
+    txn_mgr->Abort(txn1);
+    return catalog::index_oid_t(0);
+  }
+  std::shared_ptr<SqlTable> sql_table = sql_table_helper->GetSqlTable();
+  catalog::IndexHandle index_handle = catalog->GetDatabaseHandle().GetIndexHandle(txn1, db_oid);
+
+  // placeholder args
+  catalog::index_oid_t index_oid(catalog->GetNextOid());
+  auto indnatts = static_cast<int32_t>(index_attrs.size());
+  auto indnkeyatts = static_cast<int32_t>(key_attrs.size());
+  bool indisunique = unique_index;
+  bool indisprimary = false;
+  bool indisvalid = false;
+  bool indisready = true;
+  bool indislive = false;
+
+  // Intialize the index
+  Index *index = GetEmptyIndex(txn1, db_oid, index_oid, sql_table.get(), indisunique, key_attrs, catalog);
+  // Initializing the index fails
+  if (index == nullptr) {
+    txn_mgr->Abort(txn1);
+    return catalog::index_oid_t(0);
+  }
+
+  // Add IndexEntry
+  index_handle.AddEntry(txn1, index, index_oid, table_oid, indnatts, indnkeyatts, indisunique, indisprimary, indisvalid,
+                        indisready, indislive);
+
+  // initialize the building flag to false
+  auto index_id = make_index_id(db_oid, ns_oid, index_oid);
+  SetIndexBuildingFlag(index_id, false);
+
+  // Commit first transaction
+  transaction::timestamp_t commit_time = txn_mgr->Commit(txn1, TestCallbacks::EmptyCallback, nullptr);
+
+  // Wait for all transactions older than the timestamp of previous transaction commit
+  // TODO(jiaqizuo): use more efficient way to wait for all previous transactions to complete.
+  while (txn_mgr->OldestTransactionStartTime() < commit_time) {
+  }
+
+  // Start the second transaction to insert all keys into the index.
+  // The second transaction set the building flag to true in the critical section.
+  transaction::TransactionContext *txn2 =
+      txn_mgr->BeginTransactionWithAction([&]() { SetIndexBuildingFlag(index_id, true); });
+  // Change "indisready" to false and "indisvalid" to the result of populating the index in the catalog entry
+  index_handle.SetEntryColumn(txn2, index_oid, "indisready", type::TransientValueFactory::GetBoolean(false));
+  index_handle.SetEntryColumn(
+      txn2, index_oid, "indisvalid",
+      type::TransientValueFactory::GetBoolean(PopulateIndex(txn2, *sql_table, index, unique_index)));
+  // Commit the transaction
+  txn_mgr->Commit(txn2, TestCallbacks::EmptyCallback, nullptr);
+  // FIXME(xueyuanz): Delete the txns to pass the test, since the GC is disabled.
+  delete txn1;
+  delete txn2;
+  return index_oid;
+}
+
+void IndexManager::Drop(catalog::db_oid_t db_oid, catalog::namespace_oid_t ns_oid, catalog::table_oid_t table_oid,
+                        catalog::index_oid_t index_oid, const std::string &index_name,
+                        transaction::TransactionManager *txn_mgr, catalog::Catalog *catalog) {
```

Comment:
`index_name` is not used in this function, consider removing it?

Reply:
This is to be compatible with plan node arguments. Ideally this oid should be known at binding phase, but I think we can use this name as a sanity check here.